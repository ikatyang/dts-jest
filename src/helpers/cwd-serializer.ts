export const cwd_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(new RegExp(process.cwd(), 'g'), '<cwd>')),
  test: (value: any) =>
    typeof value === 'string' && value.indexOf(process.cwd()) !== -1,
};
