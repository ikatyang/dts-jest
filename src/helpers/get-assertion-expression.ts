import {
  runtime_namespace,
  IRuntimeReportParameters,
  IRuntimeSnapshotParameters,
  TriggerKind,
} from '../definitions';

export const get_assertion_expression = (kind: TriggerKind, line: number) => {
  if (kind === TriggerKind.Show) {
    const report_parameters: IRuntimeReportParameters = {line};
    const report_value = `${runtime_namespace}.report(${JSON.stringify(report_parameters)})`;
    return `console.log(${report_value})`;
  }
  const snapshot_parameters: IRuntimeSnapshotParameters = {line};
  const snapshot_value = `${runtime_namespace}.snapshot(${JSON.stringify(snapshot_parameters)})`;
  return `expect(${snapshot_value}).toMatchSnapshot()`;
};
