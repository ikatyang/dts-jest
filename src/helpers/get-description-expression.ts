import {
  runtime_namespace,
  IRuntimeDescriptionParameters,
} from '../definitions';

export const get_description_expression = (line: number) => {
  const parameters: IRuntimeDescriptionParameters = {
    line,
  };
  return `${runtime_namespace}.description(${JSON.stringify(parameters)})`;
};
