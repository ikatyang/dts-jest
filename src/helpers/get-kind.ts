import {TriggerKind} from '../definitions';

export const get_kind = (flag?: string) => {
  switch (flag) {
    case ':show':
      return TriggerKind.Show;
    case ':skip':
      return TriggerKind.Skip;
    case ':only':
      return TriggerKind.Only;
    case undefined:
      return TriggerKind.None;
    default:
      throw new Error(`Unexpected flag ${JSON.stringify(flag)}`);
  }
};
