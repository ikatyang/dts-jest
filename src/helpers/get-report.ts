import {indent} from '../utils/indent';

// tslint:disable:no-magic-numbers

export const get_report = (reporter_template: string, expression: string, snapshot: string) =>
  reporter_template
    .split(/\{\{(expression|snapshot),([0-9]+)\}\}/g)
    .map((chunk, index, chunks) =>
      (index % 3 === 0)
          ? chunk
          : (index % 3 === 1)
            ? ''
            : (chunks[index - 1] === 'snapshot')
              ? indent(snapshot, +chunk)
              : indent(expression, +chunk),
    )
    .join('');
