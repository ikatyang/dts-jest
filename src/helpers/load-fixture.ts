import * as fs from 'fs';
import * as path from 'path';
import * as _ts from 'typescript';
import { create_source_file } from '../utils/create-source-file';

export const get_fixture_filename = (id: string) =>
  path.resolve(__dirname, '../../fixtures', id);

export const load_fixture = (id: string) =>
  fs.readFileSync(get_fixture_filename(id), 'utf8');

export const load_fixture_source_file = (id: string, ts: typeof _ts) => {
  const source = load_fixture(id);
  return create_source_file(id, source, ts);
};
