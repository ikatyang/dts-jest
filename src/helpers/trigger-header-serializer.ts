import { TriggerHeader, TriggerHeaderFlags } from '../definitions';

export const trigger_header_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: TriggerHeader, serializer) => {
    const flag_literals: string[] = [];

    let { flags } = value;

    for (let base = 1; flags !== 0; flags >>= 1, base <<= 1) {
      if (flags & 1) {
        flag_literals.push(TriggerHeaderFlags[base]);
      }
    }

    return serializer({ ...value, flag_literals });
  },
  test: (value: any) =>
    typeof value === 'object' &&
    typeof value.flags === 'number' &&
    !('flag_literals' in value),
};
