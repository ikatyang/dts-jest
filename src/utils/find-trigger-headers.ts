import * as _ts from 'typescript';
import {
  trigger_header_regex,
  GroupMethod,
  TestMethod,
  TriggerGroup,
  TriggerHeader,
  TriggerHeaderFlags,
  TriggerHeaderMatchIndex,
} from '../definitions';
import { for_each_comment } from './for-each-comment';
import { get_comment_content } from './get-comment-content';
import { parse_trigger_header_flags } from './parse-trigger-header-flags';

export const find_trigger_headers = (
  source_file: _ts.SourceFile,
  ts: typeof _ts,
) => {
  const headers: TriggerHeader[] = [];

  let last_group: TriggerGroup | undefined;

  for_each_comment(
    source_file,
    (comment, scanner) => {
      const match = get_comment_content(comment).match(trigger_header_regex);

      if (match === null) {
        return;
      }

      const description = match[TriggerHeaderMatchIndex.Description];
      const flags = parse_trigger_header_flags(
        match[TriggerHeaderMatchIndex.Flags],
      );

      const start = scanner.getTokenPos();
      const { line } = source_file.getLineAndCharacterOfPosition(start);

      if (flags & TriggerHeaderFlags[':group']) {
        last_group = {
          line,
          method: get_group_method(flags),
          description: description.length === 0 ? undefined : description,
        };
      } else {
        headers.push({
          line,
          flags,
          method: get_test_method(flags),
          description: description.length === 0 ? undefined : description,
          group: last_group,
        });
      }
    },
    ts,
  );

  return headers;
};

function get_test_method(flag: TriggerHeaderFlags) {
  return flag & TriggerHeaderFlags[':only']
    ? TestMethod.Only
    : flag & TriggerHeaderFlags[':skip'] ? TestMethod.Skip : TestMethod.Test;
}

function get_group_method(flag: TriggerHeaderFlags) {
  return flag & TriggerHeaderFlags[':only']
    ? GroupMethod.Only
    : flag & TriggerHeaderFlags[':skip'] ? GroupMethod.Skip : GroupMethod.Test;
}
