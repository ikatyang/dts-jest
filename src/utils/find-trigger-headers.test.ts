import * as ts from 'typescript';
import { load_fixture_source_file } from '../helpers/load-fixture';
import { trigger_header_serializer } from '../helpers/trigger-header-serializer';
import { find_trigger_headers } from './find-trigger-headers';

expect.addSnapshotSerializer(trigger_header_serializer);

it('should return correctly', () => {
  expect(find_headers('simple')).toMatchSnapshot();
});

it('should return correctly with group', () => {
  expect(find_headers('group')).toMatchSnapshot();
});

it('should return correctly with method', () => {
  expect(find_headers('method')).toMatchSnapshot();
});

it('should do nothing with unmatched comment', () => {
  expect(find_headers('unmatched')).toHaveLength(0);
});

function find_headers(id: string) {
  return find_trigger_headers(
    load_fixture_source_file(`find-trigger-headers/${id}.ts`, ts),
    ts,
  );
}
