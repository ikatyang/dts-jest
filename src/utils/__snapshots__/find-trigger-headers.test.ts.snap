// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly 1`] = `
Array [
  Object {
    "description": undefined,
    "flag_literals": Array [],
    "flags": 0,
    "group": undefined,
    "line": 0,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":show",
    ],
    "flags": 2,
    "group": undefined,
    "line": 1,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":snap",
    ],
    "flags": 1,
    "group": undefined,
    "line": 2,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":snap",
      ":show",
    ],
    "flags": 3,
    "group": undefined,
    "line": 3,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":pass",
    ],
    "flags": 4,
    "group": undefined,
    "line": 4,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":show",
      ":pass",
    ],
    "flags": 6,
    "group": undefined,
    "line": 5,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":snap",
      ":pass",
    ],
    "flags": 5,
    "group": undefined,
    "line": 6,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":snap",
      ":show",
      ":pass",
    ],
    "flags": 7,
    "group": undefined,
    "line": 7,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":fail",
    ],
    "flags": 8,
    "group": undefined,
    "line": 8,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":show",
      ":fail",
    ],
    "flags": 10,
    "group": undefined,
    "line": 9,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":snap",
      ":fail",
    ],
    "flags": 9,
    "group": undefined,
    "line": 10,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":snap",
      ":show",
      ":fail",
    ],
    "flags": 11,
    "group": undefined,
    "line": 11,
    "method": "test",
  },
  Object {
    "description": "description",
    "flag_literals": Array [],
    "flags": 0,
    "group": undefined,
    "line": 13,
    "method": "test",
  },
  Object {
    "description": "description + flag",
    "flag_literals": Array [
      ":pass",
    ],
    "flags": 4,
    "group": undefined,
    "line": 14,
    "method": "test",
  },
]
`;

exports[`should return correctly with group 1`] = `
Array [
  Object {
    "description": "1",
    "flag_literals": Array [],
    "flags": 0,
    "group": undefined,
    "line": 0,
    "method": "test",
  },
  Object {
    "description": "2",
    "flag_literals": Array [],
    "flags": 0,
    "group": undefined,
    "line": 1,
    "method": "test",
  },
  Object {
    "description": "? 1",
    "flag_literals": Array [],
    "flags": 0,
    "group": Object {
      "description": undefined,
      "line": 2,
      "method": "describe",
    },
    "line": 3,
    "method": "test",
  },
  Object {
    "description": "? 2",
    "flag_literals": Array [],
    "flags": 0,
    "group": Object {
      "description": undefined,
      "line": 2,
      "method": "describe",
    },
    "line": 4,
    "method": "test",
  },
  Object {
    "description": "A 1",
    "flag_literals": Array [],
    "flags": 0,
    "group": Object {
      "description": "A",
      "line": 5,
      "method": "describe",
    },
    "line": 6,
    "method": "test",
  },
  Object {
    "description": "A 2",
    "flag_literals": Array [],
    "flags": 0,
    "group": Object {
      "description": "A",
      "line": 5,
      "method": "describe",
    },
    "line": 7,
    "method": "test",
  },
  Object {
    "description": "B 1",
    "flag_literals": Array [],
    "flags": 0,
    "group": Object {
      "description": "B",
      "line": 8,
      "method": "describe",
    },
    "line": 9,
    "method": "test",
  },
  Object {
    "description": "B 2",
    "flag_literals": Array [],
    "flags": 0,
    "group": Object {
      "description": "B",
      "line": 8,
      "method": "describe",
    },
    "line": 10,
    "method": "test",
  },
]
`;

exports[`should return correctly with method 1`] = `
Array [
  Object {
    "description": undefined,
    "flag_literals": Array [],
    "flags": 0,
    "group": undefined,
    "line": 0,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":only",
    ],
    "flags": 16,
    "group": undefined,
    "line": 1,
    "method": "test.only",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":skip",
    ],
    "flags": 32,
    "group": undefined,
    "line": 2,
    "method": "test.skip",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [],
    "flags": 0,
    "group": Object {
      "description": undefined,
      "line": 4,
      "method": "describe",
    },
    "line": 6,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":only",
    ],
    "flags": 16,
    "group": Object {
      "description": undefined,
      "line": 4,
      "method": "describe",
    },
    "line": 7,
    "method": "test.only",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":skip",
    ],
    "flags": 32,
    "group": Object {
      "description": undefined,
      "line": 4,
      "method": "describe",
    },
    "line": 8,
    "method": "test.skip",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [],
    "flags": 0,
    "group": Object {
      "description": undefined,
      "line": 10,
      "method": "describe.only",
    },
    "line": 12,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":only",
    ],
    "flags": 16,
    "group": Object {
      "description": undefined,
      "line": 10,
      "method": "describe.only",
    },
    "line": 13,
    "method": "test.only",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":skip",
    ],
    "flags": 32,
    "group": Object {
      "description": undefined,
      "line": 10,
      "method": "describe.only",
    },
    "line": 14,
    "method": "test.skip",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [],
    "flags": 0,
    "group": Object {
      "description": undefined,
      "line": 16,
      "method": "describe.skip",
    },
    "line": 17,
    "method": "test",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":only",
    ],
    "flags": 16,
    "group": Object {
      "description": undefined,
      "line": 16,
      "method": "describe.skip",
    },
    "line": 18,
    "method": "test.only",
  },
  Object {
    "description": undefined,
    "flag_literals": Array [
      ":skip",
    ],
    "flags": 32,
    "group": Object {
      "description": undefined,
      "line": 16,
      "method": "describe.skip",
    },
    "line": 19,
    "method": "test.skip",
  },
]
`;
